### PR TITLE
Add property handler

### PR DESCRIPTION
**Pull request by zhoonit**
Add property handler. Current validation of `setProperty` is not done
centrally, hiding bugs when invalid props are given as well as resulting
in many duplicated codes.

This property handler take care of save, load, setting default,
validation of props. So there is no need to take care of properties.
plus, this would enable automated model file save.

See also: #269, #395, #392, #391

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

**Comment by zhoonit** [comment]
This PR will be open after making some tests

File: nntrainer/include/properties.h [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+   * @brief property holder destructor
+   */
+  ~_PropertyHolder() {}
+
+  _PropertyHolder() {}
+
+  value_type get() { return data; }
+
+  void set(value_type _data) {
+    if (prop_traits<T>::is_valid(_data) == false) {
```
Comment (kparichay) [inline nntrainer/include/properties.h:179]: how about update the return value of `is_valid` to be true when valid or rename to `is_invalid`?

File: nntrainer/include/properties.h [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+      std::stringstream ss;
+      ss << "current value is not valid: " << prop_traits<T>::serialize(_data);
+      throw std::invalid_argument(ss.str().c_str());
+    }
+    force_set(_data);
+  }
+
+  void force_set(value_type _data) { data = _data; }
```
Comment (kparichay) [inline nntrainer/include/properties.h:187]: do we need to expose this?

File: nntrainer/include/properties.h [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+  static_assert(prop_traits<T>::is_property, "Only property type can be used");
+
+public:
+  typedef typename prop_traits<T>::value_type value_type;
+  typedef T prop_type;
+
+  /**
+   * @brief property holder destructor
+   */
+  ~_PropertyHolder() {}
```
Comment (kparichay) [inline nntrainer/include/properties.h:172]: lets set to default value here?

File: nntrainer/include/properties.h [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+   * @retval string reprenstation of current value
+   */
+  virtual std::string getProp() = 0;
+};
+
+/**
+ * @brief actual holder of property. This class holds pointer of actual type.
+ * Which will be residing on Properties::items;
+ */
+template <typename T> class _PropertyHolder : public PropertyHolder {
```
Comment (kparichay) [inline nntrainer/include/properties.h:162]: Just a minor thing but IMO, this class should be `PropertyHolder` and the parent class should be `_PropertyHolder`.

File: nntrainer/include/properties.h [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+  typedef T prop_type;
+
+  /**
+   * @brief property holder destructor
+   */
+  ~_PropertyHolder() {}
+
+  _PropertyHolder() {}
+
+  value_type get() { return data; }
```
Comment (kparichay) [inline nntrainer/include/properties.h:176]: add doxygen entry here and at other places.

File: nntrainer/include/properties.h [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+  typedef Color value_type;
+
+  static std::string getKey() { return "color"; }
+
+  static bool is_valid(Color c) { return c != Color::unknown; }
+
+  static Color getDefault() { return Color::unknown; }
+
+  static std::string serialize(Color c) {
+    switch (c) {
```
Comment (kparichay) [inline nntrainer/include/properties.h:107]: lets append append `key` in return of serialize and correspondingly parse it back in de-serialize

File: nntrainer/include/properties.h [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+  ~_PropertyHolder() {}
+
+  _PropertyHolder() {}
+
+  value_type get() { return data; }
+
+  void set(value_type _data) {
+    if (prop_traits<T>::is_valid(_data) == false) {
+      std::stringstream ss;
+      ss << "current value is not valid: " << prop_traits<T>::serialize(_data);
```
Comment (kparichay) [inline nntrainer/include/properties.h:181]: add key value here as well.

File: nntrainer/include/properties.h [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+public:
+  typedef typename prop_traits<T>::value_type value_type;
+  typedef T prop_type;
+
+  /**
+   * @brief property holder destructor
+   */
+  ~_PropertyHolder() {}
+
+  _PropertyHolder() {}
```
Comment (kparichay) [inline nntrainer/include/properties.h:174]: lets set to default value here?

File: nntrainer/include/properties.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/include/properties.h:179]: `is_valid` returns true already. only undefined types return `std::runtime_error`

File: nntrainer/include/properties.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/include/properties.h:187]: `force_se`t is needed to initialize data to a value that is invalid. I'll add doxygen for the function

File: nntrainer/include/properties.h (hunk 4 above)
Comment (zhoonit) [reply nntrainer/include/properties.h:162]: Could you elaborate on this? The parent class is the exposed interface and the child classes are implementations.

File: nntrainer/include/properties.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/include/properties.h]: This function is deleted.

File: nntrainer/include/properties.h (hunk 8 above)
Comment (zhoonit) [reply nntrainer/include/properties.h]: Done

File: nntrainer/include/properties.h (hunk 7 above)
Comment (zhoonit) [reply nntrainer/include/properties.h]: Done

File: nntrainer/include/properties.h (hunk 5 above)
Comment (zhoonit) [reply nntrainer/include/properties.h:185]: Done

File: nntrainer/include/properties.h [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+ *  You only have to add property struct and prop_traits.                    *
+ *                                                                           *
+ *  It is recommened to use unique key for each types.                       *
+ *  However, it is allowed to use duplicated key                             *
+ *  When those properties are mutually exclusive                             *
+ *****************************************************************************/
+
+/**< color is here for demonstration purpose. */
+
+enum class Color {
```
Comment (jijoongmoon) [inline nntrainer/include/properties.h:85]: I don't think we need to demonstrate in header. comment out this codes or have to add this in test codes.

File: nntrainer/include/properties.h [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+   * @retval string reprenstation of current value
+   */
+  virtual std::string getProp() = 0;
+};
+
+/**
+ * @brief actual holder of property. This class holds pointer of actual type.
+ * Which will be residing on Properties::items;
+ */
+template <typename T> class _PropertyHolder : public PropertyHolder {
```
Comment (jijoongmoon) [inline nntrainer/include/properties.h:162]: please not to use a name starting with ```_```

File: test/unittest/unittest_properties.cpp [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+    case value_type::UNKNOWN:
+      return "unknown";
+    default:
+      /// should not reach here
+      throw std::invalid_argument("c is not serializable");
+    }
+  }
+
+  static value_type deserialize(const std::string &val) {
+    if (val == "white")
```
Comment (jijoongmoon) [inline test/unittest/unittest_properties.cpp:133]: better to use switch and string::compare

File: test/unittest/unittest_properties.cpp [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+      return "tropical";
+    case value_type::UNKNOWN:
+      return "unknown";
+    default:
+      throw std::invalid_argument("given parameter is not serializable");
+    }
+  }
+
+  static value_type deserialize(const std::string &val) {
+    if (val == "pome")
```
Comment (jijoongmoon) [inline test/unittest/unittest_properties.cpp:83]: better to use switch and string compare?

File: nntrainer/include/properties.h [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+template <typename T> struct prop_traits {
+  static constexpr bool is_property = false; /**< check if this is property */
```
Comment (jijoongmoon) [inline nntrainer/include/properties.h:28]: struct? class? What made you to use struct? Looks like it is ok to use class. Cause struct, class mixed and make codes more complicated

**Comment by jijoongmoon** [comment]
Correct me, if I'm wrong. It looks like it is to get the properties and has feature to set the properties for developers and users.
First of all, we have to decide that we are give this feature to the users. If we are not, then it looks like overspec I guess.
And the codes looks quite complicate.
From the top to bottom, is it like
properties --> _PropertyHolder --> PropertyHolder --> proptypes --> prop_traits ?

**Comment by zhoonit** [comment]
Closing this as it is postponed to the next year

File: nntrainer/include/properties.h (hunk 4 above)
Comment (kparichay) [reply nntrainer/include/properties.h:162]: my bad. I thought _PropertyHolder was going to be exposed.

**Pull request closed without merging**